src/vmm/src/vstate/vcpu/mod.rs

```diff
@@ -0,0 +1,909 @@
+#[cfg(not(test))]
+use std::sync::Barrier;
+use std::{
+    cell::Cell,
+    fmt::{Display, Formatter},
+    io, result,
+    sync::atomic::{fence, Ordering},
+    sync::mpsc::{channel, Receiver, Sender, TryRecvError},
+    thread,
+};
+
+use crate::{
+    vmm_config::machine_config::CpuFeaturesTemplate, vstate::vm::Vm, FC_EXIT_CODE_GENERIC_ERROR,
+    FC_EXIT_CODE_OK,
+};
+use kvm_ioctls::VcpuExit;
+use logger::{error, info, Metric, METRICS};
+use seccomp::{BpfProgram, SeccompFilter};
+use utils::{
+    eventfd::EventFd,
+    signal::{register_signal_handler, sigrtmin, Killable},
+    sm::StateMachine,
+};
+
+#[cfg(target_arch = "aarch64")]
+pub(crate) mod aarch64;
+#[cfg(target_arch = "x86_64")]
+pub(crate) mod x86_64;
+
+#[cfg(target_arch = "aarch64")]
+pub(crate) use aarch64::{Error as VcpuError, *};
+#[cfg(target_arch = "x86_64")]
+pub(crate) use x86_64::{Error as VcpuError, *};
+
+/// Signal number (SIGRTMIN) used to kick Vcpus.
+pub(crate) const VCPU_RTSIG_OFFSET: i32 = 0;
+
+/// Errors associated with the wrappers over KVM ioctls.
+#[derive(Debug)]
+pub enum Error {
+    /// Error triggered by the KVM subsystem.
+    FaultyKvmExit(String),
+    /// Failed to signal Vcpu.
+    SignalVcpu(utils::errno::Error),
+    /// Kvm Exit is not handled by our implementation.
+    UnhandledKvmExit(String),
+    /// Wrapper over error triggered by some vcpu action.
+    VcpuResponse(VcpuError),
+    /// Cannot spawn a new vCPU thread.
+    VcpuSpawn(io::Error),
+    /// Cannot cleanly initialize vcpu TLS.
+    VcpuTlsInit,
+    /// Vcpu not present in TLS.
+    VcpuTlsNotPresent,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::Error::*;
+
+        match self {
+            FaultyKvmExit(ref e) => write!(f, "Received error signaling kvm exit: {}", e),
+            SignalVcpu(e) => write!(f, "Failed to signal vcpu: {}", e),
+            UnhandledKvmExit(ref e) => write!(f, "Unexpected kvm exit received: {}", e),
+            VcpuResponse(e) => write!(f, "Failed to run action on vcpu: {}", e),
+            VcpuSpawn(e) => write!(f, "Cannot spawn a new vCPU thread: {}", e),
+            VcpuTlsInit => write!(f, "Cannot clean init vcpu TLS"),
+            VcpuTlsNotPresent => write!(f, "Vcpu not present in TLS"),
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Encapsulates configuration parameters for the guest vCPUS.
+#[derive(Debug, PartialEq)]
+pub struct VcpuConfig {
+    /// Number of guest VCPUs.
+    pub vcpu_count: u8,
+    /// Enable hyperthreading in the CPUID configuration.
+    pub ht_enabled: bool,
+    /// CPUID template to use.
+    pub cpu_template: Option<CpuFeaturesTemplate>,
+}
+
+// Using this for easier explicit type-casting to help IDEs interpret the code.
+type VcpuCell = Cell<Option<*const Vcpu>>;
+
+/// A wrapper around creating and using a vcpu.
+pub struct Vcpu {
+    // Offers kvm-arch specific functionality.
+    pub kvm_vcpu: KvmVcpu,
```

Comment:
nit:
this is `pub`, everything in `KvmVcpu` is also `pub`, I prefer we didn't allow mutable access to `kvm_vcpu` from outside the `Vcpu`.
It'd be less error-prone if we did like in the case of `mmio_bus` where we have `Vcpu::set_mmio_bus()` that internally sets `vcpu.kvm_vcpu.mmio_bus`